Fix `ReorderableList` proxy animation for partial drag-back

lukemmtt opened the pull request:
_**Note:** Alongside this PR, I've also prepared [another PR](https://github.com/flutter/flutter/pull/172882) with an alternative solution involving a more substantial refactor that addresses the root cause, rather than adding more conditional logic._

## Description

This PR fixes the proxy animation bug where dragging a `ReorderableList` item downward and then back to its original position causes it to animate to the wrong location (one position too low).

## The Problem

When dragging a `ReorderableList` item downward and then back to its original position, the proxy widget briefly animates to the wrong location (one position too low) before snapping to the correct spot.

**Reproduction**: Drag any item down past at least one other item, then drag it back to where it started.

<p align="center">
  <img src="https://github.com/user-attachments/assets/d0931dff-5600-441c-8536-2c61789767d0" alt="demo2" width="250">
</p>

## Root Cause

This bug is specific to dragging an item down and then bringing it back up to nearly (but not 100% of the way ) to its original position:

1. When the item approaches its original position **from below**, `_insertIndex` becomes `item.index + 1`
    - This happens because Flutter's `ReorderableList` calculates `_insertIndex` with the dragged item still present in the list (see #24786)
2. The proxy _should_ animate to the item's original position at `item.index`
    - _But the proxy actually animates one position too low._
    - This happens because `_dragEnd` incorrectly calculates 
    `_finalDropPosition = _itemOffsetAt(_insertIndex! - 1) +
  _extentOffset(...)`
    - The `_extentOffset(...)` addition, designed for items dropping _between other items_, shifts the position down by one item's height
    - The correct calculation for "returning home from below" should be just `_itemOffsetAt(_insertIndex! - 1)`

Note that this only occurs when returning from below (`_insertIndex > item.index`). Dragging upward (in a vertical list for example) or doesn't trigger this bug.

## Existing Implementation

The existing `_dragEnd` method in `reorderable_list.dart`:

```dart
void _dragEnd(_DragInfo item) {
  setState(() {
    if (_insertIndex == item.index) {
      _finalDropPosition = _itemOffsetAt(_insertIndex!);
    } else if (_reverse) {
      if (_insertIndex! >= _items.length) {
        _finalDropPosition =
            _itemOffsetAt(_items.length - 1) - _extentOffset(item.itemExtent, _scrollDirection);
      } else {
        _finalDropPosition =
            _itemOffsetAt(_insertIndex!) +
            _extentOffset(_itemExtentAt(_insertIndex!), _scrollDirection);
      }
    } else {
      if (_insertIndex! == 0) {
        _finalDropPosition = _itemOffsetAt(0) - 
          _extentOffset(item.itemExtent, _scrollDirection);
      } else {
        _finalDropPosition = _itemOffsetAt(_insertIndex! - 1) +
          _extentOffset(_itemExtentAt(_insertIndex! - 1), _scrollDirection);
      }
    }
  });
}
```

When returning from below, the code falls through to the final else block, which incorrectly adds `_extentOffset`.

## Fix

Detect when `_insertIndex - item.index == 1` (indicating a return to original position from below) and animate to the correct position.

```dart
if (_insertIndex! - item.index == 1) {
  // Drop at the original position when item returns from below
  _finalDropPosition = _itemOffsetAt(_insertIndex! - 1);
}
```

This fix was proposed by @frankpape in https://github.com/flutter/flutter/issues/90856#issuecomment-2599368939; I've merely validated and researched the background of why the fix works, and supported it with tests. 

**_Demo of the fixed implementation:_**
<p align="center">
  <img src="https://github.com/user-attachments/assets/a53e8920-ebca-4326-abe9-3b43b34419e5" alt="fixed" width="250">
</p>

Fixes #88331
Fixes #90856
Fixes #150843

## A note about a previous PR: 

While investigating this issue, I found a PR addressing what seemed to be [the same exact issue](https://github.com/flutter/flutter/issues/150843): PR #151026; it turns out that that PR solved a _portion_ of the edge case: the case where an item is dragged down and back and slightly **overshoots** its original position when being dragged back & dropped—but that PR did not account for the presence of this bug when the dragged item slightly **undershoots** its original position on the return drag. This new PR effectively addresses the 'undershooting' case.

With this, I've added a new pair of regression tests that are identical to the [previous PR's tests](https://github.com/flutter/flutter/blob/master/packages/flutter/test/widgets/reorderable_list_test.dart#L734), except for the fact that they simulate an undershoot on the return trip (90% of the way back instead of 110% like the original tests). This definitively captures the issue, failing in the master branch and passing in this PR's branch.

Here is the specific case resolved by the [**old** PR](https://github.com/flutter/flutter/pull/151026):
<table>
  <tr>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/b0ddc745-6e9e-4f12-97da-454e2e76b06d" alt="Before" width="200"><br>
      <sub>Before</sub>
    </td>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/03e181fa-f43b-4405-b0c0-16d3465ad990" alt="After" width="200"><br>
      <sub>After</sub>
    </td>
  </tr>
</table>

Here is the additional case resolved by **this** PR:
<table>
  <tr>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/9b4bb591-aa2f-4cf0-88b8-a3ec32b0f0ac" alt="Before" width="200"><br>
      <sub>Before</sub>
    </td>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/31646e9c-78f4-4252-921f-53583193868f" alt="After" width="200"><br>
      <sub>After</sub>
    </td>
  </tr>
</table>

Two final observations worth noting:
- The fix proposed in this PR seems to **supersede** the previous PR's solution; it addresses both cases (overshooting and undershooting) even in my tests with the [original PR's changes ](https://github.com/flutter/flutter/pull/151026/files#diff-23a4bb073009d89f09084bdf5f85232de135b8f11be625e6312bb85900a90e67) reverted. Probably best to keep the old PR's code anyway to be conservative, but noteworthy.
- I also found it notable that neither this PR nor the older PR fix any issue with "reversed lists", which, in my tests, are simply not subject to this edge case as we've defined it. The regression tests added for the reverse case are thus purely precautionary.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

lukemmtt committed:
Add integration test for ReorderableList proxy animation direction bug

Tests that proxy animates upward (not downward) when returning to original position. The test:
- Drags an item away from its original position
- Returns it to the lower half of its original position (where bug occurs)
- Verifies the proxy animates upward (correct) not downward (bug)

Validates fix for #88331 and #90856

lukemmtt committed:
Fix SliverReorderableList proxy animation for partial drag-back

When dragging an item down then back up less than 100% (e.g., 90%), the proxy would animate in the wrong direction upon release. This fix properly calculates the drop position when _insertIndex - item.index == 1.

Piinks reviewed (commented):
Hey @lukemmtt! Welcome. Thank you for contributing. Rather than an integration test, this should be easier to test with a simple unit test in `packages/flutter/test/widgets/reorderable_list_test.dart` 
Can you update to test this there? I think most of CI is failing since the changes here are not quite right for setting up a new integration test, but reverting and adding a unit test should resolve those. :)

lukemmtt commented:
> Hey @lukemmtt! Welcome. Thank you for contributing. Rather than an integration test, this should be easier to test with a simple unit test in `packages/flutter/test/widgets/reorderable_list_test.dart` Can you update to test this there? I think most of CI is failing since the changes here are not quite right for setting up a new integration test, but reverting and adding a unit test should resolve those. :)

Thanks @Piinks for looking into this and for your reply!

I certainly agree that an integration test is a last resort, and I'm not surprised I set it up wrong (sorry 🥲!) but I'm afraid I was unable to create a comparable unit test despite hours of trying. The thing is, no matter what I tried, I could not manage to create a unit test that properly captured the direction of movement of the proxy as it was animating to its final position it was settling (which is the basis for the integration test). Every time I tried to 'observe' the animation during a unit test (even one frame later), I was finding that the proxy had already 'settled' into its final position, as if it wasn't animating at all. I asked AI models about it, and they just said that maybe the animation frames were being 'collapsed' in the unit test, which would make any "observing the movement of the proxy over time" an impossible task.

I'd love to be proven wrong, but I'm afraid I'm out of ideas. Open to any suggestions!

lukemmtt commented:
@Piinks I scrapped it, but I trust that if you're asking, it must be possible, so I'll be glad to investigate the existing tests and and keep trying to figure it out. I appreciate the feedback!

lukemmtt commented:
Hey @Piinks, some updates:

In looking through the [existing widget tests](https://github.com/flutter/flutter/blob/master/packages/flutter/test/widgets/reorderable_list_test.dart) for `ReorderableList`, I was surprised to find one that sounded extremely similar to the condition I'm trying to test, called ["properly animates the drop at starting position"](https://github.com/flutter/flutter/blob/fdc1b5cd29f7223213548e501c03398568997a55/packages/flutter/test/widgets/reorderable_list_test.dart#L734), added in PR #151026, which was a fix for #150843, which, in its filing video, is shown to look like this:

<img src="https://github.com/user-attachments/assets/34246064-7983-4bf2-b246-483de384142c" width="300">

...which looks visually identical to the issue(s) that I'm trying to solve in my own PR (i.e. issues https://github.com/flutter/flutter/issues/88331 and https://github.com/flutter/flutter/issues/90856). This is notable for a few reasons:

- Perhaps that issue was subtly different in some way, or perhaps it simply regressed, but it certainly persists in my own tests in the master branch, despite that PR being merged in December 2024 and [rolled into Flutter 3.27 the same month](https://docs.flutter.dev/release/release-notes/release-notes-3.27.0)).
- Moreover, that PR's [widget test](https://github.com/flutter/flutter/blob/fdc1b5cd29f7223213548e501c03398568997a55/packages/flutter/test/widgets/reorderable_list_test.dart#L734) seems to be well-crafted on the surface, but I doubt its effectiveness at detecting the [failure condition](#150843) that its PR claimed to fix, which makes me further unsure of the feasibility (or at least the proper approach) of using a widget test or unit test to reliably protect against the regression of this issue.

All that is to say: I understand that we would like to avoid an integration test if at all possible, but I remain unsure of how to make that happen effectively. That well-intentioned [widget test] (https://github.com/flutter/flutter/blob/fdc1b5cd29f7223213548e501c03398568997a55/packages/flutter/test/widgets/reorderable_list_test.dart#L734) from the earlier PR seems like a good place to revisit, but I myself am under-experienced in this department, and simply unsure how to effectively fine-tune a headless test like this.

I remain eager to learn how I can do better here. Thanks for your time and for any additional feedback!

lukemmtt committed:
Reverting integration test to explore other approaches

This reverts commit 7d0ea7f1a49a724a165c7472de5994c2c4c9c2c6.

lukemmtt committed:
Add regression tests for SliverReorderableList undershoot drag animation

Tests when dragging an item back 90% toward its original position, ensuring the proxy animates in the correct direction. Complements existing overshoot tests that test dragging back 110%.

lukemmtt commented:
@Piinks I'm pleased to report that I figured out why this issue has persisted to this day, despite seemingly having been addressed last year in PR #151026; that PR solved a _portion_ of the issue: the case where an item is dragged down and back and slightly **overshoots**  its original position when being dragged back & dropped—but that PR did not account for the presence of this bug when the dragged item slightly **undershoots** its original position on the return drag.

With this, I've added a new pair of regression tests that are identical to the [previous PR's tests](https://github.com/flutter/flutter/blob/master/packages/flutter/test/widgets/reorderable_list_test.dart#L734), except for the fact that they simulate an undershoot on the return trip (90% of the way back instead of 110% like the original tests). This definitively captures the issue, failing in the master branch and passing in this PR's branch.

Some other observations worth noting:
- The fix proposed here seems to supersede the original PR's fix; it passes all tests, even if the [original PR's changes ](https://github.com/flutter/flutter/pull/151026/files#diff-23a4bb073009d89f09084bdf5f85232de135b8f11be625e6312bb85900a90e67) are temporarily reverted. Probably best to keep the old code anyway to be conservative, but noteworthy.
- Also worth mentioning is that neither the older PR nor this one have actually resolved any problems with the "reversed list" case—my manual testing reveals that that mode has seemingly never been subject to this edge case in either form (undershoot or overshoot). I've added test coverage for it nonetheless to align with the previous PR's tests.

Piinks commented on packages/flutter/test/widgets/reorderable_list_test.dart:
```diff
@@ -798,6 +800,76 @@ void main() {
+      expect(tester.getTopLeft(find.text('item 2')), const Offset(0, 310));
+      expect(tester.getTopLeft(find.text('item 1')), const Offset(0, 300));
+
+      // Now leave the drag, it should go to index 1.
```
To clarify - I think G is off here because it is confusing item 1 for index 1. :)

Piinks reviewed (approved):
LGTM, thank you!

Piinks commented:
> Note: Alongside this PR, I've also prepared https://github.com/flutter/flutter/pull/172882 with an alternative solution involving a more substantial refactor that addresses the root cause, rather than adding more conditional logic.

LOL! I too was like, I feel like we keep adding conditional cases and was drafting a simpler flow over here. 😀 
Let's land this, and then maybe we can compare notes to refactor?

lukemmtt commented on packages/flutter/test/widgets/reorderable_list_test.dart:
```diff
@@ -798,6 +800,76 @@ void main() {
+  );
+
+  testWidgets(
+    'SliverReorderableList - properly animates the drop at starting position with reverse:true (undershoot)',
```
😯 no that is not intended—I will double check this, thanks @victorsanni

lukemmtt replied:
AH—I apologize deeply Victor—I misread your message before—I failed to scroll over to see that you were referring to the `reverse = true` test specifically.

You are correct, the reverse test passes in both `master` and in my branch.

I mentioned this in the [PR description](https://github.com/flutter/flutter/pull/172380#issuecomment-3111617979), although I admit this could have been emphasized more prominently. It's in the final section of the PR description that discusses an [earlier PR by another contributor](https://github.com/flutter/flutter/pull/151026) that only solved the issue partially:

> I also found it notable that neither this PR nor the older PR fix any issue with "reversed lists", which, in my tests, are simply not subject to this edge case as we've defined it. The regression tests added for the reverse case are thus purely precautionary.

Quite notably, in @Piinks' review of that PR, she suggested that that original PR include test coverage for the reverse case,  [comment](https://github.com/flutter/flutter/pull/151026/files#r1735058447), leading to the test that never actually failed—a pattern which I've followed for consistency.

Again, I apologize for not reading your comment more closely. I value your time and appreciate your review.

lukemmtt committed:
Merge branch 'master' into fix-reorderable-list-drag-feedback

victorsanni reviewed (approved):
LGTM - thanks for the fix!
